### PR TITLE
feat: Support pipenv

### LINKED ISSUE #2212: Support pipenv (python)
Following on from #931 

https://packaging.python.org/tutorials/managing-dependencies/
docs.pipenv.org/basics/#example-pipfile-pipfile-lock
https://realpython.com/pipenv-guide/

### PR DESCRIPTION
**Pull request by greatestape**
Add support for pipenv as a manager.

- [x] Proof of concept Pipfile extracting and updating
- [x] Add tests
- [x] Support `dev-packages`
- [x] Support non pypi sources
- [x] Support Pipfile.lock
- [x] Finalize website doc changes
- [ ] Describe requirements for running pipenv globally or via Docker

Closes #2212

### CODE REVIEW

**Comment by rarkins** [comment]
@greatestape this looks great, let me know how I can help

File: lib/manager/pipenv/update.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+module.exports = {
+  updateDependency,
+};
+
+function updateDependency(fileContent, upgrade) {
+  try {
+    logger.debug(`pipenv.updateDependency(): ${upgrade.newValue}`);
+    const pipfile = toml.parse(fileContent);
+    pipfile.packages[upgrade.depName] = upgrade.newValue;
+    return tomlify.toToml(pipfile)
```
Comment (rarkins) [inline lib/manager/pipenv/update.js]: How consistent is TOML formatting, including whitespace? Is this approach likely to produce a large diff the first time Renovate updates any dependency, if the output for `tomlify.toToml()` doesn't match with the user's previous formatting?

Comment (greatestape) [reply lib/manager/pipenv/update.js]: I have no idea. I only learned that the files are toml files and not ini files today. 

I tested it manually and it managed to only change the one line that needed changing (plus removing the trailing newline, which I can fix). It’s just one experiment with only a simple Pipfile, but it seemed promising.

Comment (rarkins) [reply lib/manager/pipenv/update.js]: That’s a good sign. Usually I go for a “keyhole surgery” approach to changing files rather than full parse/stringify because people can often be very particular about the way they want to format their files

Comment (greatestape) [reply lib/manager/pipenv/update.js]: Yeah, I'm nervous about the parse and stringify strategy too. On the other hand, I'm not super confident that I can implement a good enough ad hoc parser to enable the keyhole strategy.

What I really want is a toml parser that gives me the absolute line number. Then I could get a reliable parse of the whole file and only change the one part I care about.

Failing that, I'm going to try and add some test coverage based on more complex Pipfiles.

File: lib/manager/pipenv/extract.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      }
+      const dep = {
+        depName,
+        currentValue,
+        lineNumber,
+        purl: 'pkg:pypi/' + depName,
+        versionScheme: 'pep440',
+      };
+      logger.debug('Found a dep: ' + dep);
+      // TODO: Support other python indexes
```
Comment (rarkins) [inline lib/manager/pipenv/extract.js]: So you mean packages hosted somewhere other than pypi.org?

File: lib/manager/pipenv/index.js [hunk 3]
```diff
@@ -0,0 +1,10 @@
+const { extractDependencies } = require('./extract');
+const { updateDependency } = require('./update');
+
+const language = 'python';
+
+module.exports = {
+  extractDependencies,
+  language,
+  updateDependency,
+};
```
Comment (rarkins) [inline lib/manager/pipenv/index.js]: We need to add getArtifacts() here for updating Pipfile.lock

File: package.json [hunk 4]
```diff
@@ -114,6 +114,8 @@
     "semver-utils": "1.1.3",
     "simple-git": "1.104.0",
     "slugify": "1.3.1",
+    "toml": "^2.3.3",
+    "tomlify-j0.4": "^3.0.0",
```
Comment (rarkins) [inline package.json]: Use exact versions

**Comment by rarkins** [comment]
- [ ] Describe requirements for running pipenv globally or via Docker

**Comment by rarkins** [comment]
For this to work on the hosted app I’ll need to build a docker image that can be used to run pipenv. Here seems to be the “official” one but needs some changes, including running as non-root: 
https://github.com/pypa/pipenv/blob/master/Dockerfile

File: lib/manager/pipenv/extract.js (hunk 2 above)
Comment (rarkins) [reply lib/manager/pipenv/extract.js]: This was the best issue I found on the topic: 
https://github.com/pypa/pipenv/issues/2075

Seems like we can use the same approach in Renovate as we have for pip_requirements but we need to additionally consider how to pass to pipenv itself when updating the lock.

File: package.json (hunk 4 above)
Comment (greatestape) [reply package.json]: fixed.

**Comment by greatestape** [comment]
I don't know why the test suite is failing. The tests don't seem related to my changes and they don't fail locally. They're also not failing on the master build in travis, so I'm stumped.

ps- I have other tests that fail locally for me, but they were failing when I ran them against master too. They're all in `test/platform/git/storage`

**Comment by rarkins** [comment]
@greatestape I re-ran the test suite and it seems to be green this time. I saw that one test was a jest timeout 5000ms which unfortunately just happens when TravisCI is too slow.

**Comment by rarkins** [comment]
I've created a Dockerized `pipenv` for use by the bot: https://hub.docker.com/r/renovate/pipenv

As the image is based on Python 3.7, I thought it would have an error due to this:

```
[requires]
python_version = "3.6"
```

But it seems pipenv/python handles this backwards-compatibility automatically? Example log:

```
❯ docker run -v `pwd`:`pwd` -w `pwd` renovate/pipenv pipenv lock
Creating a virtualenv for this project…
Pipfile: /tmp/docker-pipenv/Pipfile
Using /usr/bin/python3.6 (3.6.6) to create virtualenv…
Running virtualenv with interpreter /usr/bin/python3.6
Using base prefix '/usr'
New python executable in /home/ubuntu/.local/share/virtualenvs/docker-pipenv-lJRsPNxg/bin/python3.6
Also creating executable in /home/ubuntu/.local/share/virtualenvs/docker-pipenv-lJRsPNxg/bin/python
Installing setuptools, pip, wheel...done.

Virtualenv location: /home/ubuntu/.local/share/virtualenvs/docker-pipenv-lJRsPNxg
Locking [dev-packages] dependencies…
Locking [packages] dependencies…
Updated Pipfile.lock (d8cea3)!
```

When I run it locally on OSX I do get errors:

```sh
❯ pipenv lock
Warning: the environment variable LANG is not set!
We recommend setting this in ~/.profile (or equivalent) for proper expected behavior.
Warning: Python 3.6 was not found on your system…
You can specify specific versions of Python with:
  $ pipenv --python path/to/python
```

I guess that's because I don't have python3.6 installed locally too?

In the image:

```
❯ docker run -v `pwd`:`pwd` -w `pwd` renovate/pipenv ls /usr/bin/ | grep python
python
python3
python3.6
python3.6m
python3.7
python3.7-config
python3.7m
python3.7m-config
python3m
```

Seems like only 3.6 and 3.7. Confirming that 3.5 produces errors:

```
❯ docker run -v `pwd`:`pwd` -w `pwd` renovate/pipenv pipenv lock
Warning: Python 3.5 was not found on your system…
You can specify specific versions of Python with:
  $ pipenv --python path/to/python
```

Should we support all 3.x versions in the image, or.. ?

**Comment by greatestape** [comment]
@rarkins thanks for investigating all the docker stuff. I'm not there yet. Just picking away when I get some free time.

I'm planning to get all the pipfile features supported and tested (dev-requirements, alternate sources, lock files) and then I'll start tackling the infrastructural side.

**Comment by rarkins** [comment]
@greatestape if you ever think there's something I could answer in a minute that would otherwise take you much longer to hunt down, please always ask!

**Comment by greatestape** [comment]
@rarkins Thanks for offering! I'm pretty sure I'm going to need help to figure out how to get pipenv working via docker and especially to figure out how to test it.

In the meantime, I've actually got a question about renovate behaviour: The strategy I was planning to implement for pipenv was to check `Pipfile` and `Pipfile.lock`. If it found a pinned dependency in either of those, it would update them as needed. If the requirement wasn't pinned in `Pipfile`, but it's version in `Pipfile.lock` was out of date, it would update the version in `Pipfile.lock`, leaving the `Pipfile` alone.

This is based on my understanding of how the pipenv creators intend for pipenv to be used ([this](
https://pipenv.readthedocs.io/en/latest/advanced/#pipfile-vs-setup-py) covers it pretty well). It sees to be different from how renovate handles npm lock files, so I wanted to double-check.

**Comment by rarkins** [comment]
@greatestape it's a good question about lockfile updating, but it's one I'd like to defer and support later as part of a consolidated solution for multiple package managers.

In short, I think that their vision that Pipfile dependencies "still remain somewhat vague about the exact version of a dependency to be used" is no longer the best idea once automation like Renovate exists. By keeping loose dependencies in Pipfile:

- You do not get/need any less PRs. You would still need a PR every time the lockfile needs updates anyway

- You make it far harder for developers to see at a glance what exact version of a dependency they're using. They need to dig into the `Pipfile.lock` which is far more dense and unreadable than `Pipfile`

Because pipenv if defined as a manager for apps, I can't see many reasons why all dependencies shouldn't be pinned in `Pipfile` once a tool like Renovate is in use. So therefore all direct dependencies should get updated explicitly whenever there are updates. And with any of those updates, we update the `Pipfile.lock` too.

Therefore, the only things we're missing for now are:
- People running apps and using Renovate yet want to still use ranges in `Pipfile` (they only get PR when a new out of range update is available)
- PRs for transitive dependencies within Pipfile lock, and I think the volume of those could be far too high for most anyway

Anyway I plan to solve the first use case via #1382, i.e bumping locked versions in the lockfile when they are out of date. But that can be done mostly generically without much pipfile-specific code.

**Comment by greatestape** [comment]
@rarkins That makes sense. I just read the write up in the Renovate docs that argues basically the same thing.

I do think it'd be useful if Renovate could one day create lockfile-only PRs to bump the version in the lock files directly. But I can also see how this might get hairy because at least in the case of pipenv, the lock format has been specifically designed to prevent third-party editing.

So for now I'll just try to emulate the behaviour in npm, which seems to be: bump pinned versions in the main file (`package.json`/`Pipfile`) and then update the lock file if needed.

I'm going to keep picking away at the npm manager code for my pipenv lock file implementation. I might take you up on your earlier offer if I get stuck, but I'm ok for now.

What I do need help with now is: how does renovate deal with non-pinned versions? If it's just "*", I assume it just leaves it alone. What if it's a range? Does it widen the range? Replace with a pinned version? I figured this was in the docs but I couldn't find it.

**Comment by rarkins** [comment]
> I do think it'd be useful if Renovate could one day create lockfile-only PRs to bump the version in the lock files directly. 

I agree, and would expect this within the next 1-2 months for all package managers. In general, I rarely want to impose my opinions on others so that means supporting whatever workflow they already have and would like to keep.

> If it's just "*", I assume it just leaves it alone.

Yes, that's a special case that we don't even pin even if `rangeStrategy=pin`. It's usually starred for some particular reason that would mean pinning is inappropriate.

> What if it's a range? Does it widen the range?

Its behaviour is dynamic in npm if rangeStrategy=auto, which is what preset "config:base" suggests.

- peerDependencies (not applicable to pipenv) are always widened. 
- Complex ranges (e.g. including `||`) are always widened, but I think that's also not applicable to `pipenv` because any complex type of ranges mean AND not OR. 
- So by default - unless explicitly configured to widen - we replace the range with a similar one. So this means if it was 1.x before then it would be 2.x. If it was <2.0.0 before then we'd make it <3.0.0 after 2.0.0 is released, which you could argue is "widening". 

You can see the logic here: https://github.com/renovatebot/renovate/blob/master/lib/manager/npm/range.js

And actually each package manager can define its own getRangeStrategy to return a result based on "internal" knowledge of the how the manager works. I'd consider returning "pin" for pipenv considering that these are after all apps. Users can use the preset ":preserveSemverRanges" to undo that and use "replace" instead for all.

File: lib/manager/pipenv/index.js (hunk 3 above)
Comment (greatestape) [reply lib/manager/pipenv/index.js]: I just ran across this comment again earlier today. I decided to follow your advice here. I ended up following the pattern I found in the golang manager code, because it uses getArtifacts too.

I've pushed my changes up here. They don't have any tests yet, but I wanted to get them up here so you could give them a once over to see if it makes sense.

It assumes the existence of a renovate/python image that I assume doesn't exist yet. I can probably help get that set up, though.

**Comment by rarkins** [comment]
Actually there is an image “renovate/pipenv” built and on Docker Hub already. I’m building a separate image per-tool so that in future we can have independent versioning of each for compatibility reasons.


________________________________

**Comment by greatestape** [comment]
@rarkins nice! Do you have any advice for how to set up my local environment so I can test with it?

**Comment by rarkins** [comment]
@greatestape hopefully the only thing you need is Docker installed and then via command line you specify `--binary-source=docker`. I'm happy to do the heavy lifting on that part though so stop and tell me the second you have any problems.

**Comment by rarkins** [comment]
It was hard to follow notifications earlier with the GitHub outage.

Re: lock file examples, the npm one is really complex and *not* the best to emulate. It's particularly complicated for two reasons:

1. The one "npm" package manager is actually 4 package managers in one - npm, yarn, pnpm, lerna
2. Yarn, Learna and pnpm have some *very* complex logic for linking package.json lockfiles together, i.e. the package files and lock files do not operate independently of each other

For new package mangers like `pipenv` the `composer` or `gomod` implementations are the best references.

**Comment by rarkins** [comment]
TODO: Rename `extractDependencies` to `extractPackageFile`

**Comment by greatestape** [comment]
@rarkins Sorry for the long silence. I've been in a crunch at work.

I just updated my docker code. I tried testing it but ran out of time to figure out if it's working properly.

Do you have advice for how to run the code locally so that I don't have to keep adding new requirements to my test project? :) As it is I can't figure out an easy way to just get it to forget and then re-create existing pull requests. Is there a --dry-run?

**Comment by greatestape** [comment]
If the docker stuff I did works, I think all that's left is the rename you mentioned, plus updating docs.

**Comment by greatestape** [comment]
dang. Forgot I didn't write tests for the artifact stuff.

I'll work on that next.

**Comment by rarkins** [comment]
@greatestape 

> Do you have advice for how to run the code locally so that I don't have to keep adding new requirements to my test project? :) As it is I can't figure out an easy way to just get it to forget and then re-create existing pull requests. Is there a --dry-run?

There actually is a `--dry-run` option. Let me know if you ever notice it doing anything "wet" as it's quite new.

Another technique I use is: instead of renaming/closing existing PRs to recreate them, instead just rename them with "rebase!" prefix. That will trigger Renovate to recreate/regenerate the branch's commit without creating a new PR.

File: lib/manager/pipenv/index.js (hunk 3 above)
Comment (rarkins) [reply lib/manager/pipenv/index.js]: I plan to make it `renovate/pipenv` image

**Comment by greatestape** [comment]
@rarkins thanks for doing that! I'm hoping I'll have time to get back into this tomorrow.

That weirdness you highlighted does seem to be a result of the toml de-/serialization. I'll test that Pipfile can even parse that output. It looks like it did based on the change in the lock file. Weirdly, it looks like it actually might have not been parsing it correctly in its own form, since the histogram extra wasn't there before. (could also be a bug that got fixed in the newer `pipfile`)

**Comment by rarkins** [comment]
@greatestape this was the kind of thing I was worried about with de-/serialization - we have so little control of the outputs. We might want to consider writing a bespoke updater like is done for JSON.

**Comment by rarkins** [comment]
@greatestape I rewrote the update script to do a very selective text-based replace in the file, like was already done for JSON-based package files. Now it only modifies the one line necessary! Can you test it against some of your real examples to make sure it's working as expected?

**Comment by greatestape** [comment]
@rarkins I've done some testing.

Your code's looking good.

One potential surprise: pipenv lock also updates other non-pinned requirements, so your net change for a given PR could end up having other side-effects.

I don't think there's a way around this. It's the expected behaviour for `pipenv lock`.

One other issue: My code doesn't support complex requirements like the one you linked to in your earlier example (pytest-benchmark).

I've started working on a fix, but it's a bit fussy. One solution requires teaching the the pypi/core updater part of renovate to know that the `currentVersion` value for a Pipfile dependency might be either a straight string containing the version constraints, or it might be an object with a `version` property and some other properties.

The other solution requires that the other requirements get stored in a separate field of the dependency and then restored back together before we start trying to generate new toml entries in the Pipfile updater.


I know this is a mess to read. It's hard to talk through via comment.

I'm going to be online tomorrow working on this. Let me know if you have a timeframe where we could chat in realtime. I think it would help.

**Comment by rarkins** [comment]
> One potential surprise: pipenv lock also updates other non-pinned requirements, so your net change for a given PR could end up having other side-effects.
> 
> I don't think there's a way around this. It's the expected behaviour for `pipenv lock`.

That's very unfortunate though. I was wondering if `pipenv lock` was best or something like `pipenv update foo` could isolate changes to just `foo`?

On the other hand. for an *application* - which is `pipenv`'s exclusive purpose - there is no excuse not to pin dependencies in `Pipfile`, which would reduce the possible changes to other dependencies.

> I've started working on a fix, but it's a bit fussy. One solution requires teaching the the pypi/core updater part of renovate to know that the currentVersion value for a Pipfile dependency might be either a straight string containing the version constraints, or it might be an object with a version property and some other properties.

I don't completely understand, but I think you're describing that versions can be specified simply or more complex, and we support only simple so far. If my updater code is the only problem, that's hopefully something we can overcome, ideas including:

1. If it's possible for you to detect this as part of the extract process, then add some sort of variable/flag to the dep to indicate it.
2. Or perhaps in the updater - because we are parsing to toml anyway - we do the detection there before doing the search/replace
3. If for some reason neither of those help, then we just need two different for loops to attempt the replacements and we take whichever works

> I'm going to be online tomorrow working on this. Let me know if you have a timeframe where we could chat in realtime. I think it would help.

You can try Gitter, with my github username. Message me here if it doesn't seem to be working. Alternatively, you can reach me via Skype or iMessage using the email address in my GitHub profile.

Looks like you're probably in Toronto, and I'm in Stockholm, here's the overlapping time zones: https://www.timeanddate.com/worldclock/meetingtime.html?iso=20181118&p1=239&p2=250

**Comment by rarkins** [comment]
@greatestape I'm ready to merge if you are. I've labeled it "beta" release status and disabled by default, i.e. people will need to opt-in to testing it for a while before I flip the switch to default enable it.

File: website/docs/python.md [hunk 5]
```diff
@@ -5,22 +5,40 @@ description: Python/pip dependencies support in Renovate
 
 # Python Package Manager Support
 
-Renovate supports upgrading dependencies in `pip` requirements (e.g. `requirements.txt`, `requirements.pip`) files.
+Renovate supports the following Python package managers:
+
+- `pip` (e.g. `requirements.txt`, `requirements.pip`) files
+- `pipenv` (e.g. `Pipfile`)
+- `setup.py`
```
Comment (greatestape) [inline website/docs/python.md]: @rarkins Really? Cool!

Comment (greatestape) [reply website/docs/python.md]: I mean re. `setup.py` in particular.

**Comment by greatestape** [comment]
@rarkins that's amazing! Yeah, works for me!

Thanks for doing all that, including the docs stuff! I seem to recall there was something you mentioned earlier about needing docs explaining what stuff is needed to be able to run renovate locally with Pipfile support. Is there anything outstanding there?

File: website/docs/python.md (hunk 5 above)
Comment (rarkins) [reply website/docs/python.md]: If you can test it “in production” and let me know if it works for you then that would great. Once I get some feedback then I’ll default it to enabled

**Comment by rarkins** [comment]
Actually the docs for all of them are pretty light, so I decided to defer better docs for all until later. I mean.. if you want to use pipenv then install pipenv first :)

**Comment by greatestape** [comment]
@rarkins Makes sense. I think it's good to merge.

**Comment by fahrradflucht** [comment]
@rarkins - Alright, thanks for the quick response.
I'm running the hosted Gitlab app.
